azuredevops/internal/service/approvalsandchecks/resource_check_exclusive_lock.go

```diff
@@ -0,0 +1,47 @@
+package approvalsandchecks
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/utils/pipelineschecksextras"
+)
+
+// ResourceCheckExclusiveLock schema and implementation for Exclusive Lock resource
+func ResourceCheckExclusiveLock() *schema.Resource {
+	r := genBaseCheckResource(flattenExclusiveLock, expandExclusiveLock)
+
+	r.Schema["timeout"] = &schema.Schema{
+		Type:     schema.TypeInt,
+		Optional: true,
+		Default:  43200,
+	}
+
+	return r
+}
+
+func flattenExclusiveLock(d *schema.ResourceData, check *pipelineschecksextras.CheckConfiguration, projectID string) error {
+	err := doBaseFlattening(d, check, projectID)
+	if err != nil {
+		return err
+	}
+
+	if check.Settings == nil {
```

Comment:
`doBaseFlattening` is used by other components. Should I create a custom one?

Reply:
or can I remove this check from `doBaseFlattening`?

Reply:
If I change 
```go
func expandExclusiveLock(d *schema.ResourceData) (*pipelineschecksextras.CheckConfiguration, string, error) {
	settings := make(map[string]interface{})

	return doBaseExpansion(d, approvalAndCheckType.ExclusiveLock, settings, converter.ToPtr(d.Get("timeout").(int)))
}
```

to

```go
func expandExclusiveLock(d *schema.ResourceData) (*pipelineschecksextras.CheckConfiguration, string, error) {
	return doBaseExpansion(d, approvalAndCheckType.ExclusiveLock, nil, converter.ToPtr(d.Get("timeout").(int)))
}
```

I get test errors like:
```
go test -tags all '-timeout=60s' '-parallel=4' github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/service/approvalsandchecks github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/service/build github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/service/core github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/service/git
--- FAIL: TestCheckExclusiveLock_ExpandFlatten_Roundtrip (0.00s)
    resource_check_exclusive_lock_test.go:45: 
                Error Trace:    resource_check_exclusive_lock_test.go:45
                Error:          Not equal: 
                                expected: pipelineschecksextras.CheckConfiguration{Id:(*int)(0xc6b4a0), Resource:(*pipelineschecksextras.Resource)(0xcb5a70), Type:(*pipelineschecksextras.CheckType)(0xcc24c0), Url:(*string)(nil), Links:interface {}(nil), CreatedBy:(*webapi.IdentityRef)(nil), CreatedOn:(*azuredevops.Time)(nil), Issue:(*pipelineschecksextras.CheckIssue)(nil), ModifiedBy:(*webapi.IdentityRef)(nil), ModifiedOn:(*azuredevops.Time)(nil), Settings:map[string]interface {}{}, Timeout:(*int)(0xc0000a9240)}
                                actual  : pipelineschecksextras.CheckConfiguration{Id:(*int)(0xc0002224d8), Resource:(*pipelineschecksextras.Resource)(0xc0002507e0), Type:(*pipelineschecksextras.CheckType)(0xcc24c0), Url:(*string)(nil), Links:interface {}(nil), CreatedBy:(*webapi.IdentityRef)(nil), CreatedOn:(*azuredevops.Time)(nil), Issue:(*pipelineschecksextras.CheckIssue)(nil), ModifiedBy:(*webapi.IdentityRef)(nil), ModifiedOn:(*azuredevops.Time)(nil), Settings:map[string]interface {}(nil), Timeout:(*int)(0xc0002224d0)}
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -19,4 +19,3 @@
                                  ModifiedOn: (*azuredevops.Time)(<nil>),
                                - Settings: (map[string]interface {}) {
                                - },
                                + Settings: (map[string]interface {}) <nil>,
                                  Timeout: (*int)(20000)
                Test:           TestCheckExclusiveLock_ExpandFlatten_Roundtrip
--- FAIL: TestCheckExclusiveLock_Create_DoesNotSwallowError (0.00s)
    common_check.go:145: Unexpected call to *azdosdkmocks.MockPipelineschecksextrasClient.AddCheckConfiguration([context.Background {0xc000232af0 0xc000250ca0}]) at /workspaces/terraform-provider-azuredevops/azuredevops/internal/service/approvalsandchecks/common_check.go:145 because: 
        expected call at /workspaces/terraform-provider-azuredevops/azuredevops/internal/service/approvalsandchecks/resource_check_exclusive_lock_test.go:65 doesn't match the argument at index 1.
        Got: {0xc000232af0 0xc000250ca0} (pipelineschecksextras.AddCheckConfigurationArgs)
        Want: is equal to {0xcbb960 0xcc2500} (pipelineschecksextras.AddCheckConfigurationArgs)
    controller.go:269: missing call(s) to *azdosdkmocks.MockPipelineschecksextrasClient.AddCheckConfiguration(is equal to context.Background (*context.emptyCtx), is equal to {0xcbb960 0xcc2500} (pipelineschecksextras.AddCheckConfigurationArgs)) /workspaces/terraform-provider-azuredevops/azuredevops/internal/service/approvalsandchecks/resource_check_exclusive_lock_test.go:65
    controller.go:269: aborting test due to missing call(s)
--- FAIL: TestCheckExclusiveLock_Update_DoesNotSwallowError (0.00s)
    common_check.go:195: Unexpected call to *azdosdkmocks.MockPipelineschecksextrasClient.UpdateCheckConfiguration([context.Background {0xc000232b60 0xc000251a20 0xc0002230d8}]) at /workspaces/terraform-provider-azuredevops/azuredevops/internal/service/approvalsandchecks/common_check.go:195 because: 
        expected call at /workspaces/terraform-provider-azuredevops/azuredevops/internal/service/approvalsandchecks/resource_check_exclusive_lock_test.go:148 doesn't match the argument at index 1.
        Got: {0xc000232b60 0xc000251a20 0xc0002230d8} (pipelineschecksextras.UpdateCheckConfigurationArgs)
        Want: is equal to {0xcbb960 0xcc2500 0xc6b4a0} (pipelineschecksextras.UpdateCheckConfigurationArgs)
    controller.go:269: missing call(s) to *azdosdkmocks.MockPipelineschecksextrasClient.UpdateCheckConfiguration(is equal to context.Background (*context.emptyCtx), is equal to {0xcbb960 0xcc2500 0xc6b4a0} (pipelineschecksextras.UpdateCheckConfigurationArgs)) /workspaces/terraform-provider-azuredevops/azuredevops/internal/service/approvalsandchecks/resource_check_exclusive_lock_test.go:148
    controller.go:269: aborting test due to missing call(s)
FAIL
```

Reply:
The API call for this looks as follows:
```bash
curl 'https://dev.azure.com/<org>/<resourceId>/_apis/pipelines/checks/configurations' \
--compressed -X POST \
--data-raw '{
  "type": {
    "id": "2EF31AD6-BAA0-403A-8B45-2CBC9B4E5563",
    "name": "ExclusiveLock"
  },
  "settings": {},
  "resource": {
    "type": "environment",
    "id": "<environmentId>",
    "name": "<environmentName>"
  },
  "timeout": 43200
}'
```

Reply:
So it does have a `settings` key with an empty object as value.

Reply:
`settings` not available for all type of the checks.  Exclusive lock will not return the `settings`